src/core/thread/network_data.cpp

```diff
@@ -378,19 +374,19 @@ otError NetworkData::GetNextServiceId(otNetworkDataIterator *aIterator, uint16_t
 
 bool NetworkData::ContainsOnMeshPrefixes(NetworkData &aCompare, uint16_t aRloc16)
 {
-    otNetworkDataIterator outerIterator = OT_NETWORK_DATA_ITERATOR_INIT;
-    otBorderRouterConfig  outerConfig;
-    bool                  rval = true;
+    Iterator           outerIterator = kIteratorInit;
+    OnMeshPrefixConfig outerConfig;
+    bool               rval = true;
 
-    while (aCompare.GetNextOnMeshPrefix(&outerIterator, aRloc16, &outerConfig) == OT_ERROR_NONE)
+    while (aCompare.GetNextOnMeshPrefix(outerIterator, aRloc16, outerConfig) == OT_ERROR_NONE)
     {
-        otNetworkDataIterator innerIterator = OT_NETWORK_DATA_ITERATOR_INIT;
-        otBorderRouterConfig  innerConfig;
-        otError               error;
+        Iterator           innerIterator = kIteratorInit;
+        OnMeshPrefixConfig innerConfig;
+        otError            error;
 
-        while ((error = GetNextOnMeshPrefix(&innerIterator, aRloc16, &innerConfig)) == OT_ERROR_NONE)
+        while ((error = GetNextOnMeshPrefix(innerIterator, aRloc16, innerConfig)) == OT_ERROR_NONE)
         {
-            if (memcmp(&outerConfig, &innerConfig, (sizeof(outerConfig) - sizeof(outerConfig.mRloc16))) == 0)
```

Comment:
@librasungirl, can you help identify the reason for `- sizeof(outerConfig.mRloc16)`?

Reply:
It really takes some time to recall the code several years ago. 

Now I may have thought of the intention at that time (when there might be some assumption that only one device serves as BR). However as @abtink noted, it may not take effect as the `otBorderRouterConfig` is not defined as a packed struct.

I am fine with the update here. 

Speaking of the original intention, the specification should have no rules to disallow the device which requests stable Thread Network Data to become a server, then we may still need to resolve the original concern (may in a separate PR) -- unexpected SRV_DATA.ntf from the device which acts as Server but require only stable network data.
For example,
one MTD (mode `rs`) serves as BR (`2001::/64 paosr`), due to the rules in 1.1 spec section 5.14,

> P_border_router_16 values for Border Routers that do not serve as DHCPv6 Agents are not used and set to the invalid value of 0xFFFE

This MTD server would always think its network data inconsistent and send SRV_DATA.ntf whenever there is `OT_CHANGED_THREAD_NETDATA` event in existing implementation.
https://github.com/openthread/openthread/blob/master/src/core/thread/mle.cpp#L1549

We may need to consider Server ALOCs/Border Router ALOCs etc for such device as well when determining the consistence.
And we may also need to move `Get<NetworkData::Local>().SendServerDataNotification()` when HandleAdvertisement from mle_router.cpp to mle.cpp 

Thoughts? @jwhui @abtink

Reply:
@librasungirl, I had guessed the intention had something to do with requesting only stable network data. In that case, I'm not sure how a device can determine if its information has been published to the leader. It can only determine if the prefixes exist, but not whether those prefixes were contributed by itself. That said, just determining existence may be useful in itself. Let's continue to discuss this issue separately.